src/main/jbake/content/blog/2017/07/opentracing-jaxrs.adoc

```diff
@@ -0,0 +1,136 @@
+= OpenTracing JAX-RS Instrumentation
+Pavol Loffay
+2017-7-10
+:icons: font
+:jbake-type: post
+:jbake-status: published
+:jbake-tags: blog, apm, tracing, jaeger, opentracing, jax-rs
+
+In the previous http://www.hawkular.org/blog/2017/06/9/opentracing-spring-boot.html[demo]
+we have demonstrated how to instrument Spring Boot app
+using http://opentracing.io[OpenTracing] a vendor-neutral standard for distributed tracing.
+In this article we are going to instrument JAX-RS and show how to trace business layer and add
+custom data to the trace.
+
+== Demo application
+Creating JAX-RS app from scratch might be a time consuming task, therefore in this case we are going to use
+Wildfly Swarm and use http://wildfly-swarm.io/generator/[app generator]! Choose JAX-RS and CDI dependencies and hit generate button.
+
+ifndef::env-github[]
+image::/img/blog/2017/wf-swarm-generator.png[caption="Figure 1: ", title="Wildfly Swarm generator."]
+endif::[]
+ifdef::env-github[]
+image::../../../../../assets/img/blog/2017/wf-swarm-generator.png[caption="Figure 1: ", title="Wildfly Swarm generator."]
+endif::[]
+
+The generated application contains one REST endpoint which returns hello world string.
+This endpoint is accessible on `http://localhost:8080/hello`.
+It is sufficient for our demo and in the next step we are going to add instrumentation and simple business logic.
+Do not worry at the bottom of this article are all the necessary links including source code.
+
+== Instrumentation
+To add JAX-RS OpenTracing instrumentation is very simple, just include the following dependency on
+the classpath and tracing feature will be automatically registered.
+
+----
+<dependency>
+    <groupId>io.opentracing.contrib</groupId>
+    <artifactId>opentracing-jaxrs2</artifactId>
+</dependency>
+----
+
+OpenTracing is just an API, therefore it is required to register a specific tracer instance. In this demo
+we are going to use https://github.com/uber/jaeger[Jaeger tracing system]. Tracer should be created
+and initialized only once per process, hence `ServletContextListener` is the ideal place for this task:
+
+[source,java]
+----
+@WebListener
+public class TracingContextListener implements ServletContextListener {
+
+  @Inject
+  private io.opentracing.Tracer tracer;
+
+  @Override
+  public void contextInitialized(ServletContextEvent sce) {
+    GlobalTracer.register(jaegerServer());
+  }
+
+  @Override
+  public void contextDestroyed(ServletContextEvent sce) {}
+
+  @Produces
+  @Singleton
+  public static io.opentracing.Tracer jaegerTracer() {
+    return new Configuration("wildfly-swarm", new Configuration.SamplerConfiguration(
```

Comment:
"wildfly-swarm" - think somewhere needs to explain that the tracer has been initialised with this service name - as on first inspection it is not clear that this is the purpose of this argument, and users may think this value is necessary because of using wildfly swarm.